app/models/course/experience_points_record.rb

```diff
@@ -30,4 +32,34 @@ def manually_awarded?
   def reason
     manually_awarded? ? super : specific.experience_points_display_reason
   end
+
+  private
+
+  def send_notification
+    Course::LevelNotifier.level_reached(course_user.user, level_after_update)
+  end
+
+  # Test if the course_user will reach a new level after current update.
+  def reached_new_level?
+    return false if points_awarded.nil? || !points_awarded_changed?
+
+    current_level = level_before_update
+    new_level = level_after_update
+
+    current_level.persisted? &&
+      new_level.persisted? &&
```

Comment:
Just a quick check before I propose some changes, why do you need to check if it is persisted? In case the level is deleted in the same call?